### PR TITLE
RHICOMPL-3348: Add Values to openscap parser

### PR DESCRIPTION
**Pull request by marleystipich2**
The purpose of this PR is to: 
- Parse Value definitions to get the necessary info about Values under a benchmark
Screenshot 2022-11-21 at 2 34 16 PM

- Parse the `check` / `check-export` element that can be found under Rule definitions. A Rule can have a `check-export` element under it's definition with one or more `@value-id` attributes and this would mean that the rule uses those values
Screenshot 2022-11-21 at 2 32 08 PM

### CODE REVIEW

File: lib/openscap_parser/profile.rb
Comment (marleystipich2) [inline lib/openscap_parser/profile.rb:52]: This sets up a hash that looks like `{idref1: selector1, idref2: selector2, idref3: selector3, ....}`
The key is the id of the `value` that's being refined and value of the hash is the `@selector` attribute and this will match one of the `@selector` attributes on the Value definition for the value that's being refined. This is how we find out what value the Profile wants to use.

File: lib/openscap_parser/rule.rb
Comment (marleystipich2) [inline lib/openscap_parser/rule.rb:88]: This gets a map of the ids for the values that the rule is associated with. During the import, we'll need to make sure we're associated the correct `rules` with the correct `values` because the `value` that the `rule` is referencing can be different depending on what `profile` the `rule` belongs to.

File: lib/openscap_parser/value_definition.rb
Comment (skateman) [inline lib/openscap_parser/value_definition.rb:52]: This needs some explanation...what is this and why is it needed?

**Review by skateman — CHANGES_REQUESTED** [review]

File: lib/openscap_parser/value_definition.rb
Comment (marleystipich2) [reply lib/openscap_parser/value_definition.rb:14]: Yeah, I actually used the same description method that we use in group and rule but I'll update this to use safe navigation and I don't think `|| ''` is necessary

File: lib/openscap_parser/value_definition.rb
Comment (marleystipich2) [reply lib/openscap_parser/value_definition.rb:19]: Yep, that can definitely be used here. I copied the `title` method from `rule`/`group` for this too but didn't think too much about it.

File: lib/openscap_parser/rule.rb
Comment (skateman) [reply lib/openscap_parser/rule.rb:88]: I see, okay. Please don't work on Thanksgiving :laughing:

File: lib/openscap_parser/value_definition.rb
Comment (marleystipich2) [reply lib/openscap_parser/value_definition.rb:52]: I added this comment but I don't think it was showing because I updated a small part and the file was outdated: 
```
This will give us all of the non-default values that can be included in a Value definition.
The keys of this hash will be the @selector attribute value for the non-default values. The values of this hash will be the non-default value that goes with that @selector attribute.
We'll need to import these non-default values and save their selector attributes in compliance-backend because the profiles under the benchmark can choose to use any of these values instead of the default value and the selector is how we'll know which value the profile wants to use.
```

When we're getting refined values for a profile, we can only get the `idref` and `selector` which we can use to find the actual `value`
`<refine-value idref="xccdf_org.ssgproject.content_value_var_selinux_policy_name" selector="targeted"/>`

This is why I think this method is necessary because it maps the `selectors` to their actual `values` for all of the non-default options in the Value definition.

File: lib/openscap_parser/value_definition.rb
Comment (marleystipich2) [reply lib/openscap_parser/value_definition.rb:44]: Yeah, this one is complicated. I can split them up into 2. There aren't any actual examples of complex-value in the XCCDF specs or any of the datastreams but I assumed it would look something like this: `<complex-value>604800, 2592000, 86400</complex-value>`
I need to check with the openscap team though to see exactly what that will look like but I figured we could just treat the complex value as a string because even though it is a "list" I didn't think there was a reason that we'd need to return it as an array. So using the complex-value above, the complex value would be `"604800, 2592000, 86400"` instead of `[604800, 2592000, 86400]`

File: lib/openscap_parser/value_definition.rb
Comment (marleystipich2) [reply lib/openscap_parser/value_definition.rb:27]: It's possible for there to be more than one `lower-bound` and `upper-bound` elements in a Value definition so we need to make sure we're returning the default `lower-bound`/`upper-bound` numbers here. 
The default element is the one without an empty or absent @selector, so that's what we're getting here.

But this just made me realize that when a profile uses a refined value that has the value type "number", we'll need to check if there's a non-default `lower-bound`/`upper-bound' with the same `@selector` attribute.. So we'll also need to parse non-default `lower-bound`/`upper-bound' in another method here too I think

File: lib/openscap_parser/profile.rb
Comment (marleystipich2) [reply lib/openscap_parser/profile.rb:52]: Is it okay that this method returns a hash where the key is the `idref` of a Value and the value is the `selector` associated with a value?

In the `refine-value` element in the Profile definition, the only information we have to use is the `idref` of a Value and the `selector` that is associated with the value:
`<refine-value idref="xccdf_org.ssgproject.content_value_var_selinux_policy_name" selector="targeted"/>`

So this `refined_values` method won't give us exactly what we need for this in compliance-backend:
`Profile#value_refinemens:json -> {ref_id: value, ref_id: value}`
We'll need to replace the `selectors` in the hash with the actual `value`. I don't think this would be too complicated though right?

File: lib/openscap_parser/value_definition.rb
Comment (skateman) [reply lib/openscap_parser/value_definition.rb:44]: But how does that map into our DB model then? :thinking:

File: lib/openscap_parser/profile.rb
Comment (skateman) [reply lib/openscap_parser/profile.rb:52]: I think the hash should work :+1:

**Comment by skateman** [comment]
Let's deal with refinements in a follow-up, firstly we need to unblock the rest of the work by having at least the basic stuff in the parser.

File: lib/openscap_parser/value_definition.rb
Comment (marleystipich2) [reply lib/openscap_parser/value_definition.rb:44]: I think it'd map to our DB model easily because we can just have a `value` column on the `values` table and the column type can be string so we'd store every value as a string and then store the type in a `value_type` column. I think this would be fine because we're not really doing anything with the value besides updating it sometimes so I think as long as we have the value and know what the type is we should be fine. What do you think?

Comment (skateman) [reply lib/openscap_parser/value_definition.rb:44]: I don't know, I think it should be an array :thinking: it's hard without examples, but logically that would make the most sense.

Comment (marleystipich2) [reply lib/openscap_parser/value_definition.rb:44]: I asked the openscap team what the format of complex-value is but I agree that an array would make the most sense

File: lib/openscap_parser/value_definition.rb
Comment (skateman) [reply lib/openscap_parser/value_definition.rb:27]: How woult this look like if we don't deal with refinements?

Comment (marleystipich2) [reply lib/openscap_parser/value_definition.rb:27]: This would still look like this because we still need to make sure we're getting the default lower/upper bounds which is element that has an empty `selector` attribute or no `selector` attribute

File: lib/openscap_parser/value_definition.rb
Comment (skateman) [inline lib/openscap_parser/value_definition.rb:37]: The point of [memoization](https://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/) is to avoid calculation when the method is called for the second time:
```suggestion
      @default_value ||= begin
        value_element = parsed_xml.at_xpath("value[@selector='']") || parsed_xml.at_xpath('value[not(@selector)]') || parsed_xml.xpath("value")[0]
        value_element&.text
      end
```

File: lib/openscap_parser/value_definitions.rb
Comment (skateman) [inline lib/openscap_parser/value_definitions.rb:9]: There's an included callback or something similar that does exactly what you want on compile time. Your code runs at runtime, making it slower.
```suggestion
    included do
       ...
     end
```

**Review by skateman — CHANGES_REQUESTED** [review]

File: lib/openscap_parser/value_definition.rb
Comment (marleystipich2) [reply lib/openscap_parser/value_definition.rb:37]: Ah right, this makes sense. I'll update it!

File: lib/openscap_parser/value_definitions.rb
Comment (skateman) [inline lib/openscap_parser/value_definitions.rb:9]: Please don't do this, activesupport is rails-based and this gem is not even having rails as its dependency.

File: lib/openscap_parser/value_definitions.rb
Comment (skateman) [inline lib/openscap_parser/value_definitions.rb:12]: Why does this needs to be evaluated upon inclusion?

Comment (skateman) [reply lib/openscap_parser/value_definitions.rb:12]: Nevermind, I see...copy-paste. Please go with the original (bad) pattern then :shrug:

**Review by skateman — CHANGES_REQUESTED** [review]

File: lib/openscap_parser/value_definitions.rb
Comment (marleystipich2) [reply lib/openscap_parser/value_definitions.rb]: This wasn't needed.. I fixed it

**Comment by skateman** [comment]
Looks good to me, I assume you've ran the tests locally (we really need to set up CI here). @romanblanco please take a look at it.

**Comment by skateman** [comment]
I'm okay with merging, please add an extra commit where you bump the version to 1.5.0. However, I'd hold off releasing the gem for now...

**Review by skateman — APPROVED** [review]

**Pull request merged**
